merkle/sparse_merkle_tree.go

```diff
@@ -45,13 +231,47 @@ func NewSparseMerkleTreeReader(rev int64, h MapHasher, tx storage.ReadOnlyTreeTX
 	}
 }
 
+func leafQueueSize(depths []int) int {
+	if len(depths) == 1 {
+		return 1024
+	}
+	// for higher levels make sure we've got enough space if all leaves turn out
+	// to be sub-tree futures...
+	return 1 << uint(depths[0])
+}
+
+// newLocalSubtreeWriter creates a new local go-routing based subtree worker.
+func newLocalSubtreeWriter(rev int64, prefix []byte, depths []int, tx storage.TreeTX, h TreeHasher) Subtree {
+	tree := subtreeWriter{
+		treeRevision: rev,
+		prefix:       prefix,
+		subtreeDepth: depths[0],
+		leafQueue:    make(chan func() (*indexAndHash, error), leafQueueSize(depths)),
+		root:         make(chan rootHashOrError, 1),
+		children:     make(map[string]Subtree),
+		tx:           tx,
+		treeHasher:   h,
+		getSubtree: func(p []byte) Subtree {
+			myPrefix := append(prefix, p...)
+			return newLocalSubtreeWriter(rev, myPrefix, depths[1:], tx, h)
+		},
+	}
+	// TODO(al): probably shouldn't be spawning go routines willy-nilly like
+	// this, but it'll do for now.
+	go tree.buildSubtree()
+	return &tree
+}
+
 // NewSparseMerkleTreeWriter returns a new SparseMerkleTreeWriter, which will
 // write data back into the tree at the specified revision, using the passed
 // in MapHasher to calulate/verify tree hashes, storing via tx.
 func NewSparseMerkleTreeWriter(rev int64, h MapHasher, tx storage.TreeTX) *SparseMerkleTreeWriter {
+	topSubtreeSize := 8 // must be a multiple of 8 for now.
```

Comment:
I've made it a const, but at some point I'll probably open it up to be configurable somehow - I've popped in a `TODO` for that.